File: CLDemo-OC/CLDemo-OC/Class/Base/CLNavigationController/CLNavigationController.m
//
//  NavigationController.m
//  CLDemo
//
//  Created by JmoVxia on 2016/11/17.
//  Copyright © 2016年 JmoVxia. All rights reserved.
//

#import "CLNavigationController.h"
#import "CLController.h"
#import "UIImage+CLScaleToSize.h"

@interface CLNavigationController ()

@property(nonatomic, strong) UIButton *backButton;

@end

@implementation CLNavigationController

-(UIButton *)backButton {
    if (!_backButton) {
        _backButton = [[UIButton alloc] init];
        [_backButton setImage:[[UIImage imageNamed:@"back"] imageWithTintColor:[UIColor blackColor]] forState:UIControlStateNormal];
        [_backButton addTarget:self action:@selector(backAction) forControlEvents:UIControlEventTouchUpInside];
    }
    return _backButton;
}

- (void)viewDidLoad {
    [super viewDidLoad];
    self.navigationBar.tintColor = [UIColor blackColor];
}

- (void)setNavigationBarBackgroundColor:(UIColor *)navigationBarBackgroundColor {
    self.view.backgroundColor = navigationBarBackgroundColor;
    [self.navigationBar setBackgroundImage:[UIImage imageWithColor:navigationBarBackgroundColor] forBarMetrics:UIBarMetricsDefault];
    [self.navigationBar setShadowImage:[UIImage imageWithColor:navigationBarBackgroundColor]];
    if (@available(iOS 15.0, *)) {
        UINavigationBarAppearance *appearance = [[UINavigationBarAppearance alloc] init];
        [appearance configureWithOpaqueBackground];
        appearance.backgroundColor = navigationBarBackgroundColor;
        appearance.shadowColor = [UIColor clearColor];
        self.navigationBar.standardAppearance = appearance;
        self.navigationBar.scrollEdgeAppearance = appearance;
    }
}

-(void)backAction {
    if ([self.topViewController isKindOfClass:[CLController class]]) {
        CLController *controller = (CLController *)self.topViewController;
        [controller back];
    }
}

-(void)pushViewController:(UIViewController *)viewController animated:(BOOL)animated {
    if (self.viewControllers.count > 0) {
        viewController.hidesBottomBarWhenPushed = YES;
        viewController.navigationItem.leftBarButtonItem = [[UIBarButtonItem alloc] initWithCustomView:self.backButton];
    }
    [super pushViewController:viewController animated:animated];
}

-(void)setViewControllers:(NSArray<UIViewController *> *)viewControllers animated:(BOOL)animated {
    if (viewControllers.count > 0) {
        UIViewController *lastViewController = viewControllers.lastObject;
        lastViewController.hidesBottomBarWhenPushed = YES;
        lastViewController.navigationItem.leftBarButtonItem = [[UIBarButtonItem alloc] initWithCustomView:self.backButton];
    }
    [super setViewControllers:viewControllers animated:animated];
}

-(NSArray<UIViewController *> *)popToRootViewControllerAnimated:(BOOL)animated {
    if (self.viewControllers.count > 0) {
        self.topViewController.hidesBottomBarWhenPushed = NO;
    }
    return [super popToRootViewControllerAnimated:animated];
}

-(BOOL)shouldAutorotate {
    return self.topViewController.shouldAutorotate;
}

-(UIInterfaceOrientationMask)supportedInterfaceOrientations {
    return self.topViewController.supportedInterfaceOrientations;
}

-(UIInterfaceOrientation)preferredInterfaceOrientationForPresentation {
    return self.topViewController.preferredInterfaceOrientationForPresentation;
}

-(UIStatusBarStyle)preferredStatusBarStyle {
    return self.topViewController.preferredStatusBarStyle;
}

-(BOOL)prefersStatusBarHidden {
    return self.topViewController.prefersStatusBarHidden;
}

-(UIUserInterfaceStyle)overrideUserInterfaceStyle {
    if (@available(iOS 13.0, *)) {
        return self.topViewController.overrideUserInterfaceStyle;
    } else {
        return UIUserInterfaceStyleLight;
    }
}

@end

File: CLDemo-Swift/CLDemo-Swift/Class/Base/CLNavigationController/CLNavigationController.swift
//
//  CLNavigationController.swift
//  CL
//
//  Created by Chen JmoVxia on 2020/10/16.
//  Copyright © 2020 JmoVxia. All rights reserved.
//

import UIKit


//MARK: - JmoVxia---类-属性
class CLNavigationController: UINavigationController {
    override init(nibName nibNameOrNil: String?, bundle nibBundleOrNil: Bundle?) {
        super.init(nibName: nibNameOrNil, bundle: nibBundleOrNil)
    }
    override init(rootViewController: UIViewController) {
        super.init(rootViewController: rootViewController)
    }
    required init?(coder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    deinit {
    }
    ///自定义返回响应
    var customBackActionCallback: (() -> (Bool))?
    private lazy var backItem: CLBackView = {
        let view = CLBackView()
        view.title = "    ".localized
        view.addTarget(self, action: #selector(backItemAction), for: .touchUpInside)
       return view
    }()
}
//MARK: - JmoVxia---生命周期
extension CLNavigationController {
    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(animated)
    }
    override func viewDidAppear(_ animated: Bool) {
        super.viewDidAppear(animated)
    }
    override func viewDidLoad() {
        super.viewDidLoad()
        initUI()
    }
    override func viewWillDisappear(_ animated: Bool) {
        super.viewWillDisappear(animated)
    }
    override func viewDidDisappear(_ animated: Bool) {
        super.viewDidDisappear(animated)
    }
    override func viewDidLayoutSubviews() {
        super.viewDidLayoutSubviews()
    }
}
private extension CLNavigationController {
    func initUI() {
        view.backgroundColor = .white
        UINavigationBar.appearance().tintColor = .white
    }
}
//MARK: - JmoVxia---override
extension CLNavigationController {
    override func pushViewController(_ viewController: UIViewController, animated: Bool) {
        if !viewControllers.isEmpty {
            viewController.hidesBottomBarWhenPushed = true
            viewController.navigationItem.leftBarButtonItem = UIBarButtonItem(customView: backItem)
        }
        super.pushViewController(viewController, animated: animated)
    }
    override func setViewControllers(_ viewControllers: [UIViewController], animated: Bool) {
        if !viewControllers.isEmpty {
            viewControllers.last?.hidesBottomBarWhenPushed = true
            viewControllers.last?.navigationItem.leftBarButtonItem = UIBarButtonItem(customView: backItem)
        }
        super.setViewControllers(viewControllers, animated: animated)
    }
    // 是否支持自动转屏
    override var shouldAutorotate: Bool {
        return topViewController?.shouldAutorotate ?? false
    }
    // 支持哪些屏幕方向
    override var supportedInterfaceOrientations: UIInterfaceOrientationMask {
        return topViewController?.supportedInterfaceOrientations ?? .portrait
    }
    // 默认的屏幕方向（当前ViewController必须是通过模态出来的UIViewController（模态带导航的无效）方式展现出来的，才会调用这个方法）
    override var preferredInterfaceOrientationForPresentation: UIInterfaceOrientation {
        return topViewController?.preferredInterfaceOrientationForPresentation ?? .portrait
    }
    ///状态栏样式
    override var preferredStatusBarStyle: UIStatusBarStyle {
        return topViewController?.preferredStatusBarStyle ?? .default
    }
    ///是否隐藏状态栏
    override var prefersStatusBarHidden: Bool {
        return topViewController?.prefersStatusBarHidden ?? false
    }
    @available(iOS 13.0, *)
    override var overrideUserInterfaceStyle: UIUserInterfaceStyle {
        set {
            
        }
        get {
            return topViewController?.overrideUserInterfaceStyle ?? .light
        }
    }
}
//MARK: - JmoVxia---objc
@objc extension CLNavigationController {
    func backItemAction() {
        let isCustomBack = customBackActionCallback?() ?? false
        if !isCustomBack {
            if presentingViewController != nil, viewControllers.count == 1 {
                dismiss(animated: true)
            }else {
                popViewController(animated: true)
            }
        }
    }
}
